### PR TITLE
MSI-2043: disabled source selection when disabled manage stock

### PR DESCRIPTION
**Pull request by VitaliyBoyko**
### Description
Disabled source selection algorithm when managing stock is disabled.

### Fixed Issues (if relevant)
1. magento-engcom/msi#2043: Cannot create shipment: Source item not found by source code: default and SKU: [my-sku custom option] 

#### 1
1. Create multiple sources and stocks
2. Create a product and assign it to the custom sources.
3. Disable stock managing Stores > Configuration > Catalog > Inventory > Product Stock Options > Manage Stock
4. Create an order with the product
5. Create and submit a shipment
SSA page shouldn't be shown.

#### 2
1. Create multiple sources and stocks
2. Create a product and assign it to the custom sources.
3. Disable stock managing for Product via Advanced Inventory
4. Create an order with the product
5. Create and submit a shipment
SSA page shouldn't be shown.

#### 3
1. Create multiple sources and stocks
2. Create 2 products and assign them to the custom sources.
3. Disable stock managing for 1 Product via Advanced Inventory
4. Create an order with both products
5. SSA page should be shown.
The issue with select sources for this scenario fixed in the scope of #2012.

### CODE REVIEW

**Comment by adarshkhatri** [comment]
I think we need to check this:

```
        if (!$this->stockConfiguration->getManageStock()) {
            return;
        }
```

for each product, instead of entire site.

**Comment by VitaliyBoyko** [comment]
Seems like we have to implement service which will check if the order has at least one stock manageable item. And show SSA according to result. And on SSA level we have to add the check whether item stock manageable or not. @maghamed I'm right?

**Comment by maghamed** [comment]
@VitaliyBoyko not sure that we need to introduce a dedicated service for these purposes.
Just okay to introduce dedicated class which would use InventoryConfiguration inside

**Comment by adarshkhatri** [comment]
It looks like there is check in place:

            if (!$stockItemConfiguration->isManageStock()) {
                //We don't need to Manage Stock
                continue;
            }

https://github.com/magento-engcom/msi/blob/2.3-develop/app/code/Magento/InventorySourceDeductionApi/Model/SourceDeductionService.php#L76

I think solution would be to find the list by `product id` instead of sku. I see that when creating order `Magento` adds `Custom Option` to the configurable item.

For example: 
Config item's sku: Config Item with Custom Option
Child Item: Config Item with Custom Option-Black-First

But when order created:
db

**Comment by adarshkhatri** [comment]
It looks like there is check in place:

            if (!$stockItemConfiguration->isManageStock()) {
                //We don't need to Manage Stock
                continue;
            }

https://github.com/magento-engcom/msi/blob/2.3-develop/app/code/Magento/InventorySourceDeductionApi/Model/SourceDeductionService.php#L76

I think solution would be to find the list by `product id` instead of sku. I see that when creating order `Magento` adds `Custom Option` to the configurable item, when it has custom option

**For example:** 
Config item's sku: Config Item with Custom Option
Child Item: Config Item with Custom Option-Black-First

But when order created:
db

Because of this, system will never find the product.

       //namespace Magento\InventorySourceDeductionApi\Model\GetSourceItemBySourceCodeAndSku

        $searchCriteria = $this->searchCriteriaBuilder
            ->addFilter(SourceItemInterface::SOURCE_CODE, $sourceCode) //default
            ->addFilter(SourceItemInterface::SKU, $sku) //Config Item with Custom Option-Black-First Custom Option
            ->create();
        $sourceItemsResult = $this->sourceItemRepository->getList($searchCriteria);

**I think, finding the source item by `product id` will always find the product if it exist.**

**Comment by malesh-v** [comment]
I've checked the first case, I click Ship and SSA page isn't showing(as expected) but if a click **Back** SSA page loads.

File: app/code/Magento/InventoryShippingAdminUi/Plugin/Sales/Block/Shipment/BackButtonUrlOnNewShipmentPagePlugin.php [hunk 1]
```diff
@@ -7,42 +7,59 @@
...
      * @param IsWebsiteInMultiSourceMode $isWebsiteInMultiSourceMode
+     * @param IsOrderSourceManageable $isOrderSourceManageable
      */
     public function __construct(
-        IsWebsiteInMultiSourceMode $isWebsiteInMultiSourceMode
+        IsWebsiteInMultiSourceMode $isWebsiteInMultiSourceMode,
+        IsOrderSourceManageable $isOrderSourceManageable = null
     ) {
         $this->isWebsiteInMultiSourceMode = $isWebsiteInMultiSourceMode;
+        $this->isOrderSourceManageable = $isOrderSourceManageable ??
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryShippingAdminUi/Plugin/Sales/Block/Shipment/BackButtonUrlOnNewShipmentPagePlugin.php]: Plugins are not covered with Backward Compatibility policy, constructor may be updated without additional checks

File: app/code/Magento/InventoryShippingAdminUi/Plugin/Sales/Block/Shipment/BackButtonUrlOnNewShipmentPagePlugin.php [hunk 2]
```diff
@@ -7,42 +7,59 @@
 
 namespace Magento\InventoryShippingAdminUi\Plugin\Sales\Block\Shipment;
 
+use Magento\Framework\App\ObjectManager;
+use Magento\InventoryShippingAdminUi\Model\IsOrderSourceManageable;
 use Magento\Shipping\Block\Adminhtml\Create;
 use Magento\InventoryShippingAdminUi\Model\IsWebsiteInMultiSourceMode;
 
+/**
+ * Class BackButtonUrlOnNewShipmentPagePlugin
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryShippingAdminUi/Plugin/Sales/Block/Shipment/BackButtonUrlOnNewShipmentPagePlugin.php]: This comment does not cover the purpose of the Plugin

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Comment by VitaliyBoyko** [comment]
Hi @ishakhsuvarov 
I've added the mentioned changes.
Thank you!

**Review by ishakhsuvarov — APPROVED** [review]

**Pull request merged**